### PR TITLE
interim contract log

### PR DESCRIPTION
**Pull request by mrie04**

### CODE REVIEW

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/FlatrateTermRequest/ModularFlatrateTermQuery.java [hunk 1]
```diff
@@ -49,7 +49,7 @@ public class ModularFlatrateTermQuery
 	@Nullable
 	CalendarId calendarId;
 	@Nullable
-	Timestamp dateFrom;
+	Timestamp dateFromLessOrEqual;
 	@Nullable
-	Timestamp dateTo;
+	Timestamp dateToGreaterOrEqual;
```
Comment (teosarca) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/FlatrateTermRequest/ModularFlatrateTermQuery.java:54]: now or on next issue, pls consider converting those Timestamp fields to Instant...

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/impl/FlatrateBL.java [hunk 2]
```diff
@@ -1446,9 +1451,10 @@ private I_C_Flatrate_Transition getTransitionForTerm(@NonNull final I_C_Flatrate
 	private void updateEndDate(final I_C_Flatrate_Transition transition, final I_C_Flatrate_Term term)
 	{
 		final Timestamp endDate = computeEndDate(transition, term);
-		term.setEndDate(endDate);
+		term.setEndDate(endDate == null ? term.getEndDate() : endDate);
```
Comment (teosarca) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/impl/FlatrateBL.java:1454]: IMHO this logic shall go into computeEndDate, so the returned endDate shall be already OK. wdyt?

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/impl/FlatrateDAO.java [hunk 3]
```diff
@@ -1182,14 +1180,14 @@ private ImmutableList<I_C_Flatrate_Term> retrieveByModularContractQuery(@NonNull
...
 		{
-			queryBuilder.addEqualsFilter(I_C_Flatrate_Term.COLUMNNAME_StartDate, modularFlatrateTermQuery.getDateFrom());
+			queryBuilder.addCompareFilter(I_C_Flatrate_Term.COLUMNNAME_StartDate, Operator.LESS_OR_EQUAL, modularFlatrateTermQuery.getDateFromLessOrEqual());
 		}
 
-		if (modularFlatrateTermQuery.getDateTo() != null)
+		if (modularFlatrateTermQuery.getDateToGreaterOrEqual() != null)
 		{
-			queryBuilder.addEqualsFilter(I_C_Flatrate_Term.COLUMNNAME_EndDate, modularFlatrateTermQuery.getDateTo());
+			queryBuilder.addCompareFilter(I_C_Flatrate_Term.COLUMNNAME_EndDate, Operator.GREATER_OR_EQUAL, modularFlatrateTermQuery.getDateToGreaterOrEqual());
```
Comment (teosarca) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/impl/FlatrateDAO.java:1190]: same question as above... `>=` vs `>`

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/impl/MaterialReceiptLineHandlerHelper.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+			isBillable = inOutRecord.isInterimInvoiceable();
+			className = MaterialReceiptLineInterimContractHandler.class.getName();
+		}
+		else
+		{
+			isBillable = true;
+			className = MaterialReceiptLineModularContractHandler.class.getName();
+		}
+
+		final Optional<ModularContractTypeId> modularContractTypeId = modularContractSettings.getModuleConfigs()
```
Comment (teosarca) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/impl/MaterialReceiptLineHandlerHelper.java:116]: consider using de.metas.contracts.modular.settings.ModularContractSettings#getModularContractTypeId

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/impl/FlatrateBL.java [hunk 5]
```diff
@@ -2463,4 +2471,18 @@ public Stream<FlatrateTermId> streamModularFlatrateTermIdsByQuery(@NonNull final
...
+	@Nullable
+	@Override
+	public FlatrateTermId getInterimContractIdByModularContractIdAndDate(@NonNull final FlatrateTermId modularFlatrateTermId, @NonNull final Instant date)
+	{
+		return FlatrateTermId.ofRepoIdOrNull(queryBL.createQueryBuilder(I_C_Flatrate_Term.class)
+													 .addOnlyActiveRecordsFilter()
+													 .addEqualsFilter(I_C_Flatrate_Term.COLUMNNAME_Modular_Flatrate_Term_ID, modularFlatrateTermId.getRepoId())
+													 .addEqualsFilter(I_C_Flatrate_Term.COLUMNNAME_Type_Conditions, TypeConditions.INTERIM_INVOICE)
+													 .addCompareFilter(I_C_Flatrate_Term.COLUMNNAME_StartDate, CompareQueryFilter.Operator.LESS_OR_EQUAL, date)
+													 .addCompareFilter(I_C_Flatrate_Term.COLUMNNAME_EndDate, CompareQueryFilter.Operator.GREATER_OR_EQUAL, date)
```
[the remark this answers is not in the document]
Comment (mrie04) [reply backend/de.metas.contracts/src/main/java/de/metas/contracts/impl/FlatrateBL.java:2484]: Without you would have a gap between 2 interim contracts of 1 day, where for example a receipt couldn't be assigned to any interim contract.
If one interim ends at 31.01 the next one can start at 01.02

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/interceptor/C_Flatrate_Term.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+	{
+		if (!TypeConditions.ofCode(flatrateTermRecord.getType_Conditions()).isInterimContractType())
+		{
+			return;
+		}
+
+		modularContractService.invokeWithModel(flatrateTermRecord, COMPLETED, LogEntryContractType.INTERIM);
+
+		Check.assumeNotNull(flatrateTermRecord.getEndDate(), "End Date shouldn't be null");
+		queryBL.createQueryBuilder(I_M_InOut.class)
```
Comment (teosarca) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/interceptor/C_Flatrate_Term.java:100]: can't we separate this into corresponding repositories?

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/interceptor/C_Flatrate_Term.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+			return;
+		}
+		reverseInterimReceiptLineLogsIfNeeded(flatrateTermRecord);
+		modularContractService.invokeWithModel(flatrateTermRecord, ModularContractService.ModelAction.CANCELED, LogEntryContractType.INTERIM);
+	}
+
+	private void reverseInterimReceiptLineLogsIfNeeded(@NonNull final I_C_Flatrate_Term flatrateTermRecord)
+	{
+		final Timestamp oldEndDate = InterfaceWrapperHelper.createOld(flatrateTermRecord, I_C_Flatrate_Term.class).getEndDate();
+		queryBL.createQueryBuilder(I_M_InOut.class)
```
Comment (teosarca) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/interceptor/C_Flatrate_Term.java:126]: same

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/interim/logImpl/InterimContractHandler.java [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+		}
+		final ModularContractLogEntry modularContractLogEntry = modularContractLogEntryOptional.get();
+
+		final ModularContractSettings modularContractSettings = modularContractSettingsDAO.getByFlatrateTermIdOrNull(flatrateTermId);
+		if (modularContractSettings == null)
+		{
+			return Optional.empty();
+		}
+
+		final Optional<ModularContractTypeId> modularContractTypeId = modularContractSettings.getModuleConfigs()
```
Comment (teosarca) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/interim/logImpl/InterimContractHandler.java:121]: pls use modularContractSettings's convenient method.

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/interim/logImpl/MaterialReceiptLineInterimContractHandler.java [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+	public @NonNull Stream<FlatrateTermId> streamContractIds(@NonNull final I_M_InOutLine inOutLineRecord)
+	{
+		final I_M_InOut inOutRecord = inoutDao.getById(InOutId.ofRepoId(inOutLineRecord.getM_InOut_ID()));
+		final FlatrateTermId modularFlatrateTermId = FlatrateTermId.ofRepoIdOrNull(inOutLineRecord.getC_Flatrate_Term_ID());
+		if (inOutRecord.isSOTrx() || modularFlatrateTermId == null || inOutLineRecord.getMovementQty().signum() < 0)
+		{
+			return Stream.empty();
+		}
+
+		Check.assumeNotNull(inOutRecord.getMovementDate(), "MovementDate shouldn't be null");
```
Comment (teosarca) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/interim/logImpl/MaterialReceiptLineInterimContractHandler.java:104]: when this is null? (AFAIK MovementDate is mandatory)

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/log/ModularContractLogDAO.java [hunk 10]
```diff
@@ -240,4 +253,22 @@ private static Quantity extractQty(@NonNull final I_ModCntr_Log record)
 	{
 		return Quantitys.create(record.getQty(), UomId.ofRepoId(record.getC_UOM_ID()));
 	}
+
+	public Optional<ModularContractLogEntry> getLastModularContractLog(
+			@NonNull final FlatrateTermId modularFlatrateTermId,
+			@NonNull final OrderLineId orderLineId)
+	{
+		final TableRecordReference modularRecordReference = TableRecordReference.of(I_C_OrderLine.Table_Name, orderLineId);
```
Comment (teosarca) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/log/ModularContractLogDAO.java:261]: why not using lastRecord(query)?

File: backend/de.metas.contracts/src/main/sql/postgresql/system/50-de.metas.contracts/5699910_sys_interim_contract_default_transition.sql [hunk 11]
```diff
@@ -0,0 +1,23 @@
+-- 2023-08-23T17:14:34.865196500Z
+INSERT INTO C_Flatrate_Transition (AD_Client_ID, AD_Org_ID, C_Calendar_Contract_ID, C_Flatrate_Transition_ID, Created, CreatedBy, DocAction, DocStatus, EndsWithCalendarYear, IsActive, IsAutoCompleteNewTerm, IsNotifyUserInCharge, Name, Processed, Processing, TermDuration, TermDurationUnit, TermOfNotice, TermOfNoticeUnit, Updated, UpdatedBy)
+VALUES (1000000, 1000000, 1000000, 540039, TO_TIMESTAMP('2023-08-23 19:14:34.858', 'YYYY-MM-DD HH24:MI:SS.US'), 100, 'CO', 'DR', 'N', 'Y', 'N', 'N', 'Default value for interim contract', 'N', 'N', 0, 'day', 0, 'day', TO_TIMESTAMP('2023-08-23 19:14:34.858', 'YYYY-MM-DD HH24:MI:SS.US'), 100)
```
Comment (teosarca) [inline backend/de.metas.contracts/src/main/sql/postgresql/system/50-de.metas.contracts/5699910_sys_interim_contract_default_transition.sql:3]: is it ok to have C_Calendar_Contract_ID=1000000 in migration scripts?

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/log/ModularContractLogDAO.java (hunk 10 above)
Comment (mrie04) [reply backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/log/ModularContractLogDAO.java:261]: didn't exist like that when i implemented it

**Review by teosarca — APPROVED** [review]

File: backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/contract/moduleLog/contractModuleLogPurchaseOrder.feature [hunk 12]
```diff
@@ -72,30 +72,26 @@ Feature: Modular contract log from purchase order
...
       | po_orderLine   | po_order              | module_log_product_PO   | 1000       | moduleLogConditions_PO                  |
-      | po_orderLine_2 | po_order              | module_log_product_PO   | 500        | moduleLogConditions_PO                  |
 
     When the order identified by po_order is completed
 
     And retrieve C_Flatrate_Term within 60s:
       | C_Flatrate_Term_ID.Identifier | C_Flatrate_Conditions_ID.Identifier | M_Product_ID.Identifier | OPT.C_Order_Term_ID.Identifier | OPT.C_OrderLine_Term_ID.Identifier |
       | moduleLogContract_1           | moduleLogConditions_PO              | module_log_product_PO   | po_order                       | po_orderLine                       |
-      | moduleLogContract_2           | moduleLogConditions_PO              | module_log_product_PO   | po_order                       | po_orderLine_2                     |
+
```
[the remark this answers is not in the document]
Comment (mrie04) [reply backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/contract/moduleLog/contractModuleLogPurchaseOrder.feature:81]: overlapping doesn'T allow it

File: backend/de.metas.contracts/src/main/sql/postgresql/system/50-de.metas.contracts/5699910_sys_interim_contract_default_transition.sql (hunk 11 above)
Comment (mrie04) [reply backend/de.metas.contracts/src/main/sql/postgresql/system/50-de.metas.contracts/5699910_sys_interim_contract_default_transition.sql:3]: It's our existing accounting calendar

**Pull request merged**